app/src/main/java/com/github/polybooks/core/database/DatabaseAPI.kt

```diff
@@ -0,0 +1,173 @@
+package com.github.polybooks.core.database
+
+import com.github.polybooks.core.*
+import java.util.concurrent.CompletableFuture
+
+/**
+ * Provides the API for a database. Allows to retrieve information in the database
+ * about Books, Sales.
+ * */
+interface Database {
+
+    /**
+     * Creates a new query for Books. It originally matches all books.
+     * */
+    fun queryBooks() : BookQuery
+
+    /**
+     * Create a new query for Sales. It originally matches all sales.
+     * */
+    fun querySales() : SaleQuery
+
+    /**
+     * Get all the books in the database
+     * */
+    fun listAllBooks() : CompletableFuture<List<Book>> = queryBooks().getAll()
+
+    /**
+     * Get all the sales in the database
+     * */
+    fun listAllSales() : CompletableFuture<List<Sale>> = querySales().getAll()
+
+    /**
+     * Get data about a Book from the database given it's ISBN13
+     * */
+    fun getBook(isbn13 : String) : CompletableFuture<Book>
+        = TODO("It can be implemented from the previous functions")
+
+    /**
+     * A method for getting books by batches of at most N books. The batches are indexed by ordered pages.
+     * @param numberOfBooks The maximum number of books per page
+     * @param page The index of the page
+     * @param ordering The ordering for the pages and books within the pages (see {@link BookOrdering})
+     * */
+    fun getNBooks(numberOfBooks : Int, page : Int, ordering : BookOrdering) : CompletableFuture<List<Book>>
+        = queryBooks().withOrdering(ordering).getN(numberOfBooks, page)
+
+    /**
+     * A method for getting sales by batches of at most N sales. The batches are indexed by ordered pages.
+     * @param numberOfSales The maximum number of sales per page
+     * @param page The index of the page
+     * @param ordering The ordering for the pages and sales within the pages (see {@link SaleOrdering})
+     * */
+    fun getNSales(numberOfSales : Int, page : Int, ordering : SaleOrdering) : CompletableFuture<List<Sale>>
+            = querySales().withOrdering(ordering).getN(numberOfSales, page)
+
+    //TODO need to add methods to modify the database, create entries.
+
+}
+
+/**
+ * Queries are Object that allow to make queries to a database.
+ * */
+interface Query<T> {
+
+    /**
+     * Execute the query and return all the results in a Future.
+     * */
+    fun getAll() : CompletableFuture<List<T>>
+
+    /**
+     * Get the results in batches of at most n books.
+     * */
```

Comment:
Can be changed later since very minor, but I think "books" should read "results"